### PR TITLE
Simple `MaxBoundedLen` Implementations

### PR DESCRIPTION
**Pull request by shawntabrizi**
This PR migrates some of the most important and simple pallets used in Statemint to use the `MaxBoundedLen` trait to generate `StorageInfo`.

* Assets
* Balances
* ImOnline
* Proxy
* Sudo
* Timestamp

polkadot companion: https://github.com/paritytech/polkadot/pull/3118

### CODE REVIEW

File: frame/balances/src/lib.rs [hunk 1]
```diff
@@ -826,48 +831,52 @@ impl<T: Config<I>, I: 'static> Pallet<T, I> {
...
-		}
-		// No way this can fail since we do not alter the existential balances.
-		let res = Self::mutate_account(who, |b| {
-			b.misc_frozen = Zero::zero();
-			b.fee_frozen = Zero::zero();
-			for l in locks.iter() {
-				if l.reasons == Reasons::All || l.reasons == Reasons::Misc {
-					b.misc_frozen = b.misc_frozen.max(l.amount);
+		unsafe {
+			let bounded_locks = BoundedVec::<_, T::MaxLocks>::force_from(locks.to_vec(), Some("Balances Update Locks"));
```
Comment (coriolinus) [inline frame/balances/src/lib.rs:835]: It's legal to more tightly constrain the scope of the `unsafe`, which makes it more obvious which operations were dangerous.

```suggestion
		let bounded_locks = unsafe { BoundedVec::<_, T::MaxLocks>::force_from(locks.to_vec(), Some("Balances Update Locks")) };
```

File: frame/max-encoded-len/derive/Cargo.toml [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+
+[dependencies]
+frame-support-procedural-tools = { version = "3.0.0", path = "../../support/procedural/tools" }
+proc-macro2 = "1.0.6"
+quote = "1.0.3"
+syn = { version = "1.0.58", features = ["full"] }
+
+[features]
+default = ["std"]
+std = []
```
Comment (thiolliere) [inline frame/max-encoded-len/derive/Cargo.toml:25]: I think we can remove the std feature here, as we don't use it AFAIK.

File: frame/max-encoded-len/Cargo.toml [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+	"derive",
+	"std",
+]
+derive = [
+	"max-encoded-len-derive",
+]
+std = [
+	"codec/std",
+	"max-encoded-len-derive/std",
+	"primitive-types/std",
```
Comment (thiolliere) [inline frame/max-encoded-len/Cargo.toml:34]: (note: once max encoded len is in parity-scale-codec then it should be the other way around, as codec shouldn't depend on primitive-types, but it is ok for now)

File: frame/max-encoded-len/derive/Cargo.toml (hunk 2 above)
Comment (coriolinus) [reply frame/max-encoded-len/derive/Cargo.toml:25]: We don't use it in this crate, but adding it here is how we can pass `std`-ness along to `codec` if desired.

Comment (thiolliere) [reply frame/max-encoded-len/derive/Cargo.toml:25]: codec-derive doesn't have any std feature, what does this feature means ?

File: frame/balances/src/lib.rs [hunk 4]
```diff
@@ -826,6 +831,10 @@ impl<T: Config<I>, I: 'static> Pallet<T, I> {
 
 	/// Update the account entry for `who`, given the locks.
 	fn update_locks(who: &T::AccountId, locks: &[BalanceLock<T::Balance>]) {
+		let bounded_locks = unsafe {
+			BoundedVec::<_, T::MaxLocks>::force_from(locks.to_vec(), Some("Balances Update Locks"))
+		};
+
```
Comment (thiolliere) [inline frame/balances/src/lib.rs:837]: i think we shouldn't do this. what does `force_from` means here, is it "safe" to have a vec which is bigger than its bound ? If the bound is not enforced, I think we should introduce a new struct `LightBondedVec` or something like this. Which explicitly allows to go beyond the bound without unexpected behavior.

The doc of bounded vec says:
> All internal operations ensure this bound is respected.

I open an issue for this: https://github.com/paritytech/substrate/issues/8839

Comment (shawntabrizi) [reply frame/balances/src/lib.rs:837]: @thiolliere the real solution here is to change the API of creating locks to return a result rather than always succeeding, but that is beyond the scope of this PR.

Given that we control the runtime, we actually know how many the max locks possible for a user can be, and its not attackable by a user to go past this bound unless we add pallets which would allow that.

Comment (shawntabrizi) [reply frame/balances/src/lib.rs:837]: But yeah, i will open an issue so that this api is updated

Comment (thiolliere) [reply frame/balances/src/lib.rs:837]: > Given that we control the runtime, we actually know how many the max locks possible for a user can be, and its not attackable by a user to go past this bound unless we add pallets which would allow that.

If we are sure it is correctly configured then we can do an expect with a proof here. But as long as it is not really bounded by the given configured bound, I think we should use another struct like `LightBondedVec`.

**Review by thiolliere — COMMENTED** [review]
apart from the comment on lock bounded vec it looks good to me. https://github.com/paritytech/substrate/pull/8793#discussion_r633584065

Introducing LightBoundedVec should be enough in my opinion

File: frame/proxy/src/lib.rs [hunk 5]
```diff
@@ -752,32 +772,3 @@ impl<T: Config> Pallet<T> {
 		Self::deposit_event(Event::ProxyExecuted(e.map(|_| ()).map_err(|e| e.error)));
 	}
 }
-
-/// Migration utilities for upgrading the Proxy pallet between its different versions.
-pub mod migration {
```
Comment (apopiak) [inline frame/proxy/src/lib.rs:757]: Tagged as `E1-runtimemigration` because you remove the migration here.
Would prefer to remove it in a separate PR so it's easier to isolate.

Comment (shawntabrizi) [reply frame/proxy/src/lib.rs:757]: have to remove it because the changes to bounded vec are affected here.

File: frame/proxy/src/lib.rs [hunk 6]
```diff
@@ -558,7 +567,13 @@ pub mod pallet {
 		_,
 		Twox64Concat,
 		T::AccountId,
-		(Vec<ProxyDefinition<T::AccountId, T::ProxyType, T::BlockNumber>>, BalanceOf<T>),
+		(
+			BoundedVec<
```
Comment (thiolliere) [inline frame/proxy/src/lib.rs:571]: due to the usage of bounded vec here we should make sure that the limit is not exceeded before migration.
Also any modification of `T::MaxProxies` should ensure the on chain items don't exceed the limit.

If these consideration are too strong, we should move to WeakBoundedVec.

File: frame/proxy/src/lib.rs [hunk 7]
```diff
@@ -569,7 +584,13 @@ pub mod pallet {
 		_,
 		Twox64Concat,
 		T::AccountId,
-		(Vec<Announcement<T::AccountId, CallHashOf<T>, T::BlockNumber>>, BalanceOf<T>),
+		(
+			BoundedVec<
```
Comment (thiolliere) [inline frame/proxy/src/lib.rs:588]: due to the usage of bounded vec here we should make sure that the limit is not exceeded before migration.
Also any modification of `T::MaxProxies` should ensure the on chain items don't exceed the limit.

If these consideration are too strong, we should move to WeakBoundedVec.

File: frame/assets/src/lib.rs [hunk 8]
```diff
@@ -902,8 +910,14 @@ pub mod pallet {
 		) -> DispatchResult {
 			let origin = ensure_signed(origin)?;
 
-			ensure!(name.len() <= T::StringLimit::get() as usize, Error::<T, I>::BadMetadata);
-			ensure!(symbol.len() <= T::StringLimit::get() as usize, Error::<T, I>::BadMetadata);
+			let bounded_name: BoundedVec<u8, T::StringLimit> = name
```
Comment (thiolliere) [inline frame/assets/src/lib.rs:913]: due to the usage of bounded vec here we should make sure that the limit is not exceeded before migration.
Also any modification of `T::StringLimit` should ensure the on chain items don't exceed the limit.

If these consideration are too strong, we should move to WeakBoundedVec.

Also we could use BoundedVec for the call argument directly instead of Vec + conversion here. But I think this can be done later, as UI needs to handle the new type.

File: max-encoded-len/derive/tools/src/lib.rs [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+	} else {
+		let mod_name = generate_hidden_includes_mod_name(unique_id);
+		quote::quote!( self::#mod_name::hidden_include )
+	}
+}
+
+/// Generate the crate access for the crate using 2018 syntax.
+///
+/// for `frame-support` output will for example be `frame_support`.
+pub fn generate_crate_access_2018(def_crate: &str) -> Result<syn::Ident, Error> {
```
Comment (coriolinus) [inline max-encoded-len/derive/tools/src/lib.rs:52]: We could probably move this function directly into `max-encoded-len/derive/src/lib.rs` and then rm both the `temp-procedural-tools` and `test-procedural-tools-derive` crates entirely.

File: max-encoded-len/src/lib.rs [hunk 10]
```diff
@@ -1,6 +1,6 @@
 // This file is part of Substrate.
 
-// Copyright (C) 2019-2021 Parity Technologies (UK) Ltd.
+// Copyright (C) 2021 Parity Technologies (UK) Ltd.
```
Comment (coriolinus) [inline max-encoded-len/src/lib.rs:3]: Was this change intentional?

**Review by coriolinus — APPROVED** [review]
Several nits--and I authored much of the proc-macro stuff of this PR, so there should definitely be more eyes there--but I think that fundamentally this all looks good.

File: max-encoded-len/src/lib.rs (hunk 10 above)
Comment (shawntabrizi) [reply max-encoded-len/src/lib.rs:3]: yeah, since this file was only made this year, the C should start this year?

File: max-encoded-len/derive/tools/src/lib.rs (hunk 9 above)
Comment (coriolinus) [reply max-encoded-len/derive/tools/src/lib.rs:52]: I mean sure, but even so it would be a bit embarrassing to accidentally move two entire extra crates into `parity-scale-codec` when we really only need these 15 lines of code.

File: frame/support/src/lib.rs [hunk 11]
```diff
@@ -2341,3 +2341,7 @@ pub mod pallet_prelude {
 /// * use the newest nightly possible.
 ///
 pub use frame_support_procedural::pallet;
+
+/// The `max_encoded_len` module contains the `MaxEncodedLen` trait and derive macro, which is
+/// useful for computing upper bounds on storage size.
+pub use max_encoded_len;
```
[the remark this answers is not in the document]
Comment (shawntabrizi) [reply frame/support/src/lib.rs:2347]: Honestly at this point, this re-export saves a lot of cargo.toml changes. Lets keep it in until we refactor for real.

File: frame/balances/src/lib.rs [hunk 12]
```diff
@@ -424,7 +425,9 @@ pub mod pallet {
 		Blake2_128Concat,
 		T::AccountId,
 		AccountData<T::Balance>,
-		ValueQuery
+		ValueQuery,
+		GetDefault,
+		ConstU32<300_000>,
```
Comment (thiolliere) [inline frame/balances/src/lib.rs:430]: I guess this will need to be configurable in the future, because chains can have very different number of account

File: frame/balances/src/lib.rs [hunk 13]
```diff
@@ -435,8 +438,10 @@ pub mod pallet {
 		_,
 		Blake2_128Concat,
 		T::AccountId,
-		Vec<BalanceLock<T::Balance>>,
-		ValueQuery
+		WeakBoundedVec<BalanceLock<T::Balance>, T::MaxLocks>,
+		ValueQuery,
+		GetDefault,
+		ConstU32<300_000>,
```
Comment (thiolliere) [inline frame/balances/src/lib.rs:444]: same, it will need to be configurable

File: frame/assets/src/lib.rs [hunk 14]
```diff
@@ -255,8 +261,10 @@ pub mod pallet {
 		_,
 		Blake2_128Concat,
 		T::AssetId,
-		AssetMetadata<DepositBalanceOf<T, I>>,
+		AssetMetadata<DepositBalanceOf<T, I>, BoundedVec<u8, T::StringLimit>>,
```
Comment (thiolliere) [inline frame/assets/src/lib.rs:264]: I think in the future (or now) we should say how the associated types can be modified.
For instance here "T::StringLimit" is used for stored item, thus changing the limit requires a migration of those items.

**Review by thiolliere — APPROVED** [review]

**Comment by shawntabrizi** [comment]
bot merge

File: frame/assets/src/lib.rs (hunk 14 above)
Comment (shawntabrizi) [reply frame/assets/src/lib.rs:264]: Only when the value decreases. Increases should be backwards compatible.

**Comment by thiolliere** [comment]
> I see this has E1-runtimemigration. Can you help me understand what that implies for downstream chains? For example, Moonbeam uses some of these pallets (Balances, Proxy, Sudo, and Timestamp). Do we need a storage migration? If so what would it be?

this is the potential migration needed I found:
* in pallet asset: we now use a BoundedVec in the type `AssetMetadata` which is stored in `Metadata` storage. If a value in `Metadata` storage contains a value with a name or a symbol of length more than `Config::StringLimit::get()`, then the value must be migrated in some way (removed, or updated with a name and symbol of length less or equal to `Config::StringLimit::get()`.
This can happen if the runtime has set a limit for `StringLimit` and then decreased the limit in another runtime upgrade, while keeping some value which were accepted by the previous limit in storage.

* in proxy pallet: we now use a BoundedVec in the storage `Proxies`. if a value in `Proxries` contains more a vec of length more than `Config::MaxProxies::get()` then it must be migrated.
This can happen if the runtime has set a limit for `MaxProxies` and then decreased the limit in another runtime upgrade, while keeping some value which were accepted by the previous limit in storage.

File: frame/proxy/src/lib.rs [hunk 15]
```diff
@@ -749,32 +769,3 @@ impl<T: Config> Pallet<T> {
...
-					.map(|(a, t)| ProxyDefinition {
-						delegate: a,
-						proxy_type: t,
-						delay: Zero::zero(),
-					})
-					.collect::<Vec<_>>(),
-				deposit,
-			))
-		);
-		T::BlockWeights::get().max_block
```
Comment (apopiak) [inline frame/proxy/src/lib.rs:778]: Note that this removal is another reason this is tagged as E1-runtimemigration
